lib/recog/fingerprint.rb

```diff
@@ -43,7 +47,33 @@ def initialize(xml)
   # @return [Hash,nil] Keys will be host, service, and os attributes
   def match(match_string)
     # match_string.force_encoding('BINARY') if match_string
-    match_data = @regex.match(match_string)
+    begin
+      match_data = @regex.match(match_string)
+    rescue Encoding::CompatibilityError => e
+      begin
+        # Replace invalid UTF-8 characters with spaces, just as DAP does.
+        encoded_str = match_string.encode("UTF-8", :invalid => :replace, :undef => :replace, :replace => '')
+        match_data = @regex.match(encoded_str)
+      rescue Exception => e
```

Comment:
Actually, `match_string` is relevant to both since it is the original data that was being processed. Either we generated an exception on line 51 during the original parse, 55 during encoding, or on line 56 while running regex against the newly encoded `match_string`.  While  the error message is a bit verbose, with its output we should be able to reconstruct any issue without having the source data.

I've cleaned it up a touch and added a bit more context.  The output now looks like:

```
Exception while running regex against match_string
#<Encoding::CompatibilityError: incompatible encoding regexp match (UTF-8 regexp with ASCII-8BIT string)>
Length:   14
Encoding: ASCII-8BIT
Problematic data:
/R?{??\ͥ
Raw bytes:
"\x18/\x03R\eH\x8D{\xD6\x1C\xC8\\\xCD\xA5"
```